source/MaterialXRuntime/Private/PrvTypeDef.cpp

```diff
@@ -0,0 +1,369 @@
+//
+// TM & (c) 2017 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#include <MaterialXRuntime/Private/PrvTypeDef.h>
+
+#include <MaterialXRuntime/RtTypeDef.h>
+
+#include <MaterialXCore/Util.h>
+
+#include <sstream>
+
+namespace MaterialX
+{
+
+namespace
```

Comment:
I decided to not use the Value class from core for storing values in the runtime. Mainly because those values are all of varying size and always heap allocated. I wanted to make the memory layout of nodes in the runtime as efficient as possible w.r.t cache coherence. So all ports on a node is stored in an single thunk of memory. With the exception of values that are larger than 16 bytes and don't fit naturally into an RtValue. This is the case for matrices and arrays (later), which needs to be heap allocated and have their pointer stored instead. But in the other 99% of the cases the value fits into the nodes data storage, and heap allocations and scattered memory access / cache misses are minimized.